webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/MeshMappingHelper.scala

```diff
@@ -23,55 +23,59 @@ trait MeshMappingHelper {
       mappingNameForMeshFile: Option[String],
       omitMissing: Boolean, // If true, failing lookups in the agglomerate file will just return empty list.
       token: Option[String])(implicit ec: ExecutionContext): Fox[List[Long]] =
-    targetMappingName match {
-      case None =>
+    (targetMappingName, editableMappingTracingId) match {
+      case (None, None) =>
         // No mapping selected, assume id matches meshfile
         Fox.successful(List(agglomerateId))
-      case Some(mappingName) if mappingNameForMeshFile.contains(mappingName) =>
+      case (Some(mappingName), None) if mappingNameForMeshFile.contains(mappingName) =>
         // Mapping selected, but meshfile has the same mapping name in its metadata, assume id matches meshfile
         Fox.successful(List(agglomerateId))
-      case Some(mappingName) =>
+      case (Some(mappingName), None) =>
         // Mapping selected, but meshfile does not have matching mapping name in its metadata,
         // assume agglomerate id, fetch oversegmentation segment ids for it
-        val agglomerateFileKey = AgglomerateFileKey(
-          organizationName,
-          datasetName,
-          dataLayerName,
-          mappingName
-        )
-        editableMappingTracingId match {
-          case Some(tracingId) =>
-            for {
-              tracingstoreUri <- dsRemoteWebknossosClient.getTracingstoreUri
-              segmentIdsResult <- dsRemoteTracingstoreClient.getEditableMappingSegmentIdsForAgglomerate(tracingstoreUri,
-                                                                                                        tracingId,
-                                                                                                        agglomerateId,
-                                                                                                        token)
-              segmentIds <- if (segmentIdsResult.agglomerateIdIsPresent)
-                Fox.successful(segmentIdsResult.segmentIds)
-              else
-                for {
-                  agglomerateService <- binaryDataServiceHolder.binaryDataService.agglomerateServiceOpt.toFox
-                  localSegmentIds <- agglomerateService.segmentIdsForAgglomerateId(
-                    agglomerateFileKey,
-                    agglomerateId
-                  )
-                } yield localSegmentIds
-            } yield segmentIds
-          case _ =>
+        for {
+          agglomerateService <- binaryDataServiceHolder.binaryDataService.agglomerateServiceOpt.toFox
+          segmentIdsBox <- agglomerateService
+            .segmentIdsForAgglomerateId(
+              AgglomerateFileKey(
+                organizationName,
+                datasetName,
+                dataLayerName,
+                mappingName
+              ),
```

Comment:
You have a duplication of this `AgglomerateFileKey` constructer (see lines 69ff.). Prior to these changes there was only one definition of the constructor and its result was stored in a val agglomerateFileKey.

=> A minor suggestion from my side would be to restore this and only have code that constructs the `AgglomerateFileKey` once :)

Reply:
Thanks for the suggestion! Yes, I have already considered this. However, I ultimately decided against it, since the only ways that I saw to deduplicate this (without completely changing the logic again) would be either to construct the AgglomerateFileKey above the whole match statement, but that doesn’t seem right, since in most cases it won’t be used, or to extract it to another function, but that doesn’t seem right, because its call would look exactly the same as the direct constructor call.

So I tend to think in this case breaking the DRY-rule is less bad than the alternatives :)

Reply:
Sure, sounds like a good reason to me :+1: